### ISSUE TITLE
Type checking missing some polymorphic instantiation errors

### ISSUE THREAD

**Issue by nickbattle** [overturetool/overture#562]
The following specification should have a type error:

```
functions
    f[@T]: () -> nat
    f() == g[@T](p[@T]);   -- p[@T] is not compatible with the parameter of g[@T]

    p[@T]: set of @T -> nat
    p(a) == card a;

    g[@T]: (@T -> nat) -> nat
    g(a) == a(1);
```

Note that "f[nat]()" will call "g[nat](p[nat])", but that "p[nat]" is  of type "set of nat -> nat", whereas the parameter to g requires "nat -> nat". Currently, no error is given.

**Comment by nickbattle** [comment]
Now fixed in ncb/development.

**Comment by nickbattle** [comment]
This is more complicated than it first appears. In the example given, if `@T` is something like (nat | set of nat), then the `set of @T` in p could be a set of nat, and the `@T` receiving that value can accept it. So in fact the specification is valid, by possible semantics. I think it is generally the case that any aggregate type (seq, set, map etc) involving `@T` is compatible with a bare `@T` since the type could be of the form (nat | aggregate of nat).

**Comment by nickbattle** [comment]
I finally got round to looking at this. It's quite tricky to spot the cases where there *might* be an issue, but I've added a new warning which will appear if an example (such as the one above) is given. The warning means that you *might* be right, but it's slightly unusual and if it is not what you intended then you will have a very hard time sorting out what the problem is!

The spec above now gives:
```
Warning 5031: Type @T must be a union in 'DEFAULT' (test.vdm) at line 3:15
```
That is on the instantiation argument for `g[@T]`. It means that the only way that g can be instantiated is if it is a union of types that satisfies both the `p[@T]` and the `g[@T]`.

In our experience (hi Leo), these warnings should only occur if the user has made a mistake.
